File: go/090-bank-account/bank_account.go
package account

import "sync"

type Account struct {
	balance int64
	mu      sync.Mutex
	open    bool
}

func Open(amount int64) *Account {
	if amount < 0 {
		return nil
	}

	return &Account{
		balance: amount,
		open:    true,
	}
}

func (a *Account) Balance() (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()

	if a.open == false {
		return 0, false
	}

	return a.balance, true
}

func (a *Account) Deposit(amount int64) (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()

	if a.open == false {
		return 0, false
	}

	if a.balance < -amount {
		return 0, false
	}

	a.balance += amount
	return a.balance, true
}

func (a *Account) Close() (int64, bool) {
	a.mu.Lock()
	defer a.mu.Unlock()

	if !a.open {
		return 0, false
	}

	a.open = false
	payout := a.balance
	a.balance = 0
	return payout, true
}

File: ruby/120-bank-account/bank_account.rb
class BankAccount
  def initialize
    @balance = 0
    @open = false
  end

  def open
    raise ArgumentError, "You can't open an already open account" if @open
    @open = true
    @balance = 0
  end

  def close
    raise ArgumentError, "You can't close an already closed account" unless @open
    @open = false
  end

  def balance
    raise ArgumentError, "You can't check the balance of a closed account" unless @open
    @balance
  end

  def deposit(amount)
    raise ArgumentError, "You can't deposit money into a closed account" unless @open
    raise ArgumentError, "You can't deposit a negative amount" if amount < 0
    @balance += amount
  end

  def withdraw(amount)
    raise ArgumentError, "You can't withdraw money into a closed account" unless @open
    raise ArgumentError, "You can't withdraw a negative amount" if amount < 0
    raise ArgumentError, "You can't withdraw more than you have" if amount > @balance
    @balance -= amount
  end
end
